internal/exec/stages/disks/luks.go

```diff
@@ -0,0 +1,211 @@
+	Thumbprint string `json:"thp,omitempty"`
+}
+
+type Pin struct {
+	Tpm  bool   `json:"tpm"`
+	Tang []Tang `json:"tang,omitempty"`
+}
+
+func (p Pin) MarshalJSON() ([]byte, error) {
+	if p.Tpm {
+		return json.Marshal(&struct {
+			Tang []Tang   `json:"tang,omitempty"`
+			Tpm  struct{} `json:"tpm2"`
+		}{
+			Tang: p.Tang,
+			Tpm:  struct{}{},
+		})
+	}
+	return json.Marshal(&struct {
+		Tang []Tang `json:"tang"`
+	}{
+		Tang: p.Tang,
+	})
+}
+
+type Clevis struct {
+	Pins      Pin `json:"pins"`
+	Threshold int `json:"t"`
+}
+
+// Initially tested generating keyfiles via dd'ing to a file from /dev/urandom
+// however while cryptsetup had no problem with these keyfiles clevis seemed to
+// die on them while keyfiles generated via openssl rand -hex would work...
+func randHex(length int) (string, error) {
+	bytes := make([]byte, length)
+	if _, err := rand.Read(bytes); err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(bytes), nil
+}
+
+func (s *stage) closeLuks(name string) func() error {
+	return func() error {
+		if _, err := s.Logger.LogCmd(
+			exec.Command(distro.CryptsetupCmd(), "luksClose", name),
+			"closing luks device %v", name,
+		); err != nil {
+			return fmt.Errorf("closing luks device: %v", err)
+		}
+		return nil
+	}
+}
+
+func (s *stage) createLuks(config types.Config) error {
+	if len(config.Storage.Luks) == 0 {
+		return nil
+	}
+
+	for _, luks := range config.Storage.Luks {
+		// track whether Ignition creates the KeyFile
+		// so that it can be removed creation
+		var ignitionCreatedKeyFile bool
+		// create keyfile inside of tmpfs, it will be copied to the
+		// sysroot by ignition-copy-keyfiles
+		os.MkdirAll(filepath.Join("/run", execUtil.LuksKeyFilePath), 0644)
+		keyFilePath := filepath.Join("/run", execUtil.LuksKeyFilePath, luks.Name)
+		if luks.KeyFile == nil || *luks.KeyFile == "" {
+			// create a keyfile
+			key, err := randHex(4096)
+			if err != nil {
+				return fmt.Errorf("generating keyfile: %v", err)
+			}
+			if err := ioutil.WriteFile(keyFilePath, []byte(key), 0400); err != nil {
+				return fmt.Errorf("creating keyfile: %v", err)
+			}
+			ignitionCreatedKeyFile = true
+		} else {
+			if err := ioutil.WriteFile(keyFilePath, []byte(*luks.KeyFile), 0400); err != nil {
+				return fmt.Errorf("writing keyfile: %v", err)
+			}
+		}
+
+		args := []string{
+			"luksFormat",
+			"--type", "luks2",
+			"--key-file", keyFilePath,
+		}
+
+		if !util.NilOrEmpty(luks.Hash) {
+			args = append(args, "--hash", *luks.Hash)
+		}
+
+		if !util.NilOrEmpty(luks.Label) {
+			args = append(args, "--label", *luks.Label)
+		}
+
+		if !util.NilOrEmpty(luks.UUID) {
+			args = append(args, "--uuid", *luks.UUID)
+		}
+
+		if !util.NilOrEmpty(luks.Cipher) {
+			args = append(args, "--cipher", *luks.Cipher)
+		}
+
+		if len(luks.Options) > 0 {
+			// golang's a really great language...
+			for _, option := range luks.Options {
+				args = append(args, string(option))
+			}
+		}
+
+		args = append(args, luks.Device)
```

Comment:
Are we sanity-checking that no two LUKS objects are targeting the same device? Actually, I guess this should apply across `disks`/`raid`/`filesystems`/`luks`.

Reply:
For just the individual sections I believe we're covered (iirc unless the field is marked to ignore duplicates they are required to be unique).

That said I don't believe we currently have any coverage that the combination of sections targeting devices do not contain duplicates.